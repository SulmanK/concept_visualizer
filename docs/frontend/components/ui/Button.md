# Button Component

The `Button` component provides a standardized button element for user interactions throughout the application.

## Overview

This component is a custom Tailwind CSS-based button with consistent styling, variants, and accessibility features. It simplifies implementing buttons with a consistent look and feel across the application, including hover animations and reduced motion support.

## Usage

```tsx
import { Button } from 'components/ui/Button';

// Basic usage
<Button onClick={handleClick}>
  Click Me
</Button>

// Primary variant
<Button variant="primary" onClick={handleClick}>
  Primary Action
</Button>

// Secondary variant
<Button variant="secondary" onClick={handleCancel}>
  Secondary Action
</Button>

// Outline variant
<Button variant="outline" onClick={handleCancel}>
  Outlined Button
</Button>

// Pill-shaped button (fully rounded)
<Button pill onClick={handleAction}>
  Pill Button
</Button>
```

## Props

| Prop        | Type                                               | Default     | Description                           |
| ----------- | -------------------------------------------------- | ----------- | ------------------------------------- |
| `children`  | `React.ReactNode`                                  | -           | Button content                        |
| `variant`   | `'primary' \| 'secondary' \| 'outline' \| 'ghost'` | `'primary'` | Button style variant                  |
| `size`      | `'sm' \| 'md' \| 'lg'`                             | `'md'`      | Button size                           |
| `pill`      | `boolean`                                          | `false`     | Whether to use fully rounded corners  |
| `animated`  | `boolean`                                          | `true`      | Whether to use hover/press animations |
| `className` | `string`                                           | `''`        | Additional CSS class to apply         |
| `type`      | `'button' \| 'submit' \| 'reset'`                  | `'button'`  | HTML button type                      |
| `disabled`  | `boolean`                                          | `false`     | Whether button is disabled            |
| `onClick`   | `(e: React.MouseEvent<HTMLButtonElement>) => void` | -           | Click handler                         |

## Implementation Details

```tsx
import React, { useState } from "react";
import { usePrefersReducedMotion } from "../../hooks";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Button variant
   */
  variant?: "primary" | "secondary" | "outline" | "ghost";

  /**
   * Button size
   */
  size?: "sm" | "md" | "lg";

  /**
   * Whether to use a pill shape (fully rounded)
   */
  pill?: boolean;

  /**
   * Button type attribute
   */
  type?: "button" | "submit" | "reset";

  /**
   * Additional class names
   */
  className?: string;

  /**
   * Button children
   */
  children: React.ReactNode;

  /**
   * Whether to add subtle hover animation
   * @default true
   */
  animated?: boolean;
}

/**
 * Button component with different variants and sizes
 */
export const Button: React.FC<ButtonProps> = ({
  variant = "primary",
  size = "md",
  pill = false,
  type = "button",
  className = "",
  children,
  animated = true,
  onClick,
  ...props
}) => {
  const [isPressed, setIsPressed] = useState(false);
  const prefersReducedMotion = usePrefersReducedMotion();

  // Base classes for all buttons
  const baseClasses =
    "inline-flex items-center justify-center font-medium focus:outline-none focus:ring-2 focus:ring-primary/30 focus:ring-offset-2 disabled:opacity-60 disabled:cursor-not-allowed";

  // Transition classes (skip if reduced motion is preferred)
  const transitionClasses = prefersReducedMotion
    ? ""
    : "transition-all duration-200";

  // Define variant-specific classes
  const variantClasses = {
    primary:
      "bg-gradient-to-r from-primary to-primary-dark text-white shadow-modern hover:shadow-modern-hover hover:brightness-105",
    secondary:
      "bg-gradient-to-r from-secondary to-secondary-dark text-white shadow-modern hover:shadow-modern-hover hover:brightness-105",
    outline:
      "border border-indigo-300 text-indigo-700 bg-white hover:bg-indigo-50 hover:text-primary-dark",
    ghost: "text-indigo-600 hover:text-indigo-800 hover:bg-indigo-50",
  };

  // Size classes
  const sizeClasses = {
    sm: "text-xs px-2.5 py-1",
    md: "text-sm px-4 py-2",
    lg: "text-base px-6 py-3",
  };

  // Border radius based on pill prop
  const roundedClasses = pill ? "rounded-full" : "rounded-lg";

  // Animation classes for pressed state
  const animationClasses =
    animated && !prefersReducedMotion
      ? isPressed
        ? "transform scale-95"
        : "transform scale-100 hover:scale-[1.02]"
      : "";

  // Combine all classes
  const buttonClasses = `${baseClasses} ${transitionClasses} ${variantClasses[variant]} ${sizeClasses[size]} ${roundedClasses} ${animationClasses} ${className}`;

  // Handle click with animation
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (animated && !prefersReducedMotion && !props.disabled) {
      setIsPressed(true);

      // Reset the pressed state after animation
      setTimeout(() => {
        setIsPressed(false);
      }, 150);
    }

    // Call the original onClick handler
    if (onClick) {
      onClick(e);
    }
  };

  return (
    <button
      type={type}
      className={buttonClasses}
      onClick={handleClick}
      {...props}
    >
      {children}
    </button>
  );
};
```

## Features

- **Tailwind CSS Styling**: Uses Tailwind utility classes for consistent styling
- **Multiple Variants**: Primary, secondary, outline, and ghost variants
- **Motion Sensitivity**: Respects user motion preferences
- **Animations**: Subtle scale animations on hover and click
- **Gradient Backgrounds**: Primary and secondary variants use gradient backgrounds
- **Size Options**: Small, medium, and large sizes
- **Shape Options**: Standard rounded or pill shape
- **Accessible**: Includes focus rings and respects disabled state
- **Modern Shadow**: Custom shadow effects for depth

## Usage Patterns

### Primary Action Button

```tsx
<Button variant="primary" onClick={handleSubmit} type="submit">
  Submit Form
</Button>
```

### Secondary Action Button

```tsx
<Button variant="secondary" onClick={handleCancel}>
  Cancel
</Button>
```

### Outline Button

```tsx
<Button variant="outline" onClick={handleReset}>
  Reset Form
</Button>
```

### Ghost Button

```tsx
<Button variant="ghost" onClick={handleSkip}>
  Skip this step
</Button>
```

### Pill-Shaped Button

```tsx
<Button variant="primary" pill onClick={handleAction}>
  Action Button
</Button>
```

### Different Sizes

```tsx
<>
  <Button variant="primary" size="sm">
    Small
  </Button>
  <Button variant="primary" size="md">
    Medium
  </Button>
  <Button variant="primary" size="lg">
    Large
  </Button>
</>
```

### Disabled Button

```tsx
<Button variant="primary" disabled>
  Can't Click Me
</Button>
```

## Best Practices

1. Use primary variant for the main action in a form or view
2. Use secondary variant for secondary actions
3. Use outline or ghost variants for less prominent actions
4. Consider using pill shape for action buttons that should stand out
5. Use appropriate button sizes based on importance and context
6. Keep button text concise and action-oriented
7. Ensure sufficient contrast between button color and text for accessibility
8. Group related buttons together with consistent styling
9. Leverage the className prop to add custom styles when needed
10. Set the animated prop to false for utility buttons that don't need animation
