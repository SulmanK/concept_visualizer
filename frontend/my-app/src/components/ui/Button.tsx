import React, { useState } from 'react';
import { usePrefersReducedMotion } from '../../hooks';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Button variant
   */
  variant?: 'primary' | 'secondary' | 'outline' | 'ghost';
  
  /**
   * Button size
   */
  size?: 'sm' | 'md' | 'lg';
  
  /**
   * Whether to use a pill shape (fully rounded)
   */
  pill?: boolean;
  
  /**
   * Button type attribute
   */
  type?: 'button' | 'submit' | 'reset';
  
  /**
   * Additional class names
   */
  className?: string;
  
  /**
   * Button children
   */
  children: React.ReactNode;
  
  /**
   * Whether to add subtle hover animation
   * @default true
   */
  animated?: boolean;
}

/**
 * Button component with different variants and sizes
 */
export const Button: React.FC<ButtonProps> = ({
  variant = 'primary',
  size = 'md',
  pill = false,
  type = 'button',
  className = '',
  children,
  animated = true,
  onClick,
  ...props
}) => {
  const [isPressed, setIsPressed] = useState(false);
  const prefersReducedMotion = usePrefersReducedMotion();
  
  // Base classes for all buttons
  const baseClasses = 'inline-flex items-center justify-center font-medium focus:outline-none focus:ring-2 focus:ring-primary/30 focus:ring-offset-2 disabled:opacity-60 disabled:cursor-not-allowed';
  
  // Transition classes (skip if reduced motion is preferred)
  const transitionClasses = prefersReducedMotion ? '' : 'transition-all duration-200';
  
  // Define variant-specific classes
  const variantClasses = {
    primary: 'bg-gradient-to-r from-primary to-primary-dark text-white shadow-modern hover:shadow-modern-hover hover:brightness-105',
    secondary: 'bg-gradient-to-r from-secondary to-secondary-dark text-white shadow-modern hover:shadow-modern-hover hover:brightness-105',
    outline: 'border border-indigo-300 text-indigo-700 bg-white hover:bg-indigo-50 hover:text-primary-dark',
    ghost: 'text-indigo-600 hover:text-indigo-800 hover:bg-indigo-50',
  };
  
  // Size classes
  const sizeClasses = {
    sm: 'text-xs px-2.5 py-1',
    md: 'text-sm px-4 py-2',
    lg: 'text-base px-6 py-3',
  };
  
  // Border radius based on pill prop
  const roundedClasses = pill ? 'rounded-full' : 'rounded-lg';
  
  // Animation classes for pressed state
  const animationClasses = animated && !prefersReducedMotion
    ? isPressed
      ? 'transform scale-95' 
      : 'transform scale-100 hover:scale-[1.02]'
    : '';
  
  // Combine all classes
  const buttonClasses = `${baseClasses} ${transitionClasses} ${variantClasses[variant]} ${sizeClasses[size]} ${roundedClasses} ${animationClasses} ${className}`;
  
  // Handle click with animation
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (animated && !prefersReducedMotion && !props.disabled) {
      setIsPressed(true);
      
      // Reset the pressed state after animation
      setTimeout(() => {
        setIsPressed(false);
      }, 150);
    }
    
    // Call the original onClick handler
    if (onClick) {
      onClick(e);
    }
  };
  
  return (
    <button
      type={type}
      className={buttonClasses}
      onClick={handleClick}
      {...props}
    >
      {children}
    </button>
  );
}; 