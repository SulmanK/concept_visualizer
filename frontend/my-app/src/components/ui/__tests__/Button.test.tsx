import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Button } from '../Button';

describe('Button Component', () => {
  // Basic rendering tests
  test('renders button with children', () => {
    render(<Button>Click me</Button>);
    
    const button = screen.getByRole('button', { name: /click me/i });
    expect(button).toBeInTheDocument();
  });
  
  test('renders button with primary variant by default', () => {
    render(<Button>Default Button</Button>);
    
    const button = screen.getByRole('button');
    expect(button.className).toContain('bg-gradient-primary');
  });
  
  // Variant tests
  test.each([
    ['primary', 'bg-gradient-primary'],
    ['secondary', 'bg-gradient-secondary'],
    ['outline', 'border border-indigo-300'],
    ['ghost', 'text-indigo-600'],
  ])('renders %s variant correctly', (variant, expectedClass) => {
    render(<Button variant={variant as any}>Button</Button>);
    
    const button = screen.getByRole('button');
    expect(button.className).toContain(expectedClass);
  });
  
  // Size tests
  test.each([
    ['sm', 'text-xs'],
    ['md', 'text-sm'],
    ['lg', 'text-base'],
  ])('renders %s size correctly', (size, expectedClass) => {
    render(<Button size={size as any}>Button</Button>);
    
    const button = screen.getByRole('button');
    expect(button.className).toContain(expectedClass);
  });
  
  // Pill shape test
  test('renders pill shape when pill is true', () => {
    render(<Button pill>Pill Button</Button>);
    
    const button = screen.getByRole('button');
    expect(button.className).toContain('rounded-full');
  });
  
  // Event tests
  test('calls onClick when button is clicked', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Clickable Button</Button>);
    
    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
  
  test('does not call onClick when button is disabled', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick} disabled>Disabled Button</Button>);
    
    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).not.toHaveBeenCalled();
  });
  
  // Snapshot tests
  describe('Snapshots', () => {
    test('default button snapshot', () => {
      const { container } = render(<Button>Default Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('primary variant button snapshot', () => {
      const { container } = render(<Button variant="primary">Primary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('secondary variant button snapshot', () => {
      const { container } = render(<Button variant="secondary">Secondary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('outline variant button snapshot', () => {
      const { container } = render(<Button variant="outline">Outline Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('ghost variant button snapshot', () => {
      const { container } = render(<Button variant="ghost">Ghost Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('disabled button snapshot', () => {
      const { container } = render(<Button disabled>Disabled Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('pill button snapshot', () => {
      const { container } = render(<Button pill>Pill Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
}); 