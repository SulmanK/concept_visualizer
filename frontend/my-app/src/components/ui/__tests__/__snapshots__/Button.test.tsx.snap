// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Snapshots > default button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-primary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-lg "
  type="button"
>
  Default Button
</button>
`;

exports[`Button Component > Snapshots > disabled button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-primary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-lg "
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Button Component > Snapshots > ghost variant button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed text-indigo-600 hover:text-indigo-800 hover:bg-indigo-50 text-sm px-4 py-2 rounded-lg "
  type="button"
>
  Ghost Button
</button>
`;

exports[`Button Component > Snapshots > outline variant button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed border border-indigo-300 text-indigo-700 bg-white hover:bg-indigo-50 hover:text-primary-dark text-sm px-4 py-2 rounded-lg "
  type="button"
>
  Outline Button
</button>
`;

exports[`Button Component > Snapshots > pill button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-primary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-full "
  type="button"
>
  Pill Button
</button>
`;

exports[`Button Component > Snapshots > primary variant button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-primary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-lg "
  type="button"
>
  Primary Button
</button>
`;

exports[`Button Component > Snapshots > secondary variant button snapshot 1`] = `
<button
  class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-secondary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-lg "
  type="button"
>
  Secondary Button
</button>
`;
